File: libs/ui-canvas/src/main/kotlin/mono/html/canvas/canvas/InteractionCanvasViewController.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.html.canvas.canvas

import kotlin.math.PI
import kotlin.math.abs
import mono.common.exhaustive
import mono.graphics.geo.MousePointer
import mono.graphics.geo.Point
import mono.lifecycle.LifecycleOwner
import mono.livedata.LiveData
import mono.livedata.map
import mono.shapebound.InteractionBound
import mono.shapebound.InteractionPoint
import mono.shapebound.LineInteractionBound
import mono.shapebound.ScalableInteractionBound
import mono.ui.theme.ThemeColor
import org.w3c.dom.HIGH
import org.w3c.dom.HTMLCanvasElement
import org.w3c.dom.ImageSmoothingQuality
import org.w3c.dom.Path2D

/**
 * A canvas view controller to render the interaction indicators.
 */
internal class InteractionCanvasViewController(
    lifecycleOwner: LifecycleOwner,
    canvas: HTMLCanvasElement,
    drawingInfoLiveData: LiveData<DrawingInfoController.DrawingInfo>,
    mousePointerLiveData: LiveData<MousePointer>
) : BaseCanvasViewController(canvas) {

    var interactionBounds: List<InteractionBound> = emptyList()

    private var isMouseMoving: Boolean = false

    init {
        drawingInfoLiveData.observe(lifecycleOwner, listener = ::setDrawingInfo)
        mousePointerLiveData.map { it is MousePointer.Drag }.observe(lifecycleOwner) {
            isMouseMoving = it
        }
        context.imageSmoothingQuality = ImageSmoothingQuality.HIGH
    }

    override fun drawInternal() {
        for (bound in interactionBounds) {
            when (bound) {
                is ScalableInteractionBound -> drawScalableInteractionBound(bound)
                is LineInteractionBound -> drawLineInteractionBound(bound)
            }.exhaustive
        }
    }

    private fun drawScalableInteractionBound(bound: ScalableInteractionBound) {
        val leftPx = drawingInfo.toXPx(bound.left)
        val rightPx = drawingInfo.toXPx(bound.right)
        val topPx = drawingInfo.toYPx(bound.top)
        val bottomPx = drawingInfo.toYPx(bound.bottom)
        val path = Path2D().apply {
            moveTo(leftPx, topPx)
            lineTo(rightPx, topPx)
            lineTo(rightPx, bottomPx)
            lineTo(leftPx, bottomPx)
            closePath()
        }
        context.strokeStyle = ThemeColor.SelectionBoundStroke.colorCode
        context.lineWidth = 1.0
        context.stroke(path)

        if (isMouseMoving) {
            return
        }

        bound.interactionPoints.draw(ThemeColor.SelectionDotStroke, ThemeColor.SelectionDotFill)
    }

    private fun drawLineInteractionBound(bound: LineInteractionBound) {
        if (isMouseMoving) {
            return
        }

        bound.interactionPoints.draw(
            ThemeColor.SelectionDotStroke,
            ThemeColor.SelectionDotFill
        )
        // Draw a small dot inside the start dot to make it simpler to distinguish start and end
        // dots.
        bound.interactionPoints.take(1)
            .draw(ThemeColor.SelectionDotStroke, ThemeColor.SelectionDotStroke, 1.0)
    }

    private fun List<InteractionPoint>.draw(
        strokeColor: ThemeColor,
        fillColor: ThemeColor,
        dotRadius: Double = DOT_RADIUS
    ) {
        val dotPath = Path2D()
        context.beginPath()
        for (point in this) {
            dotPath.addDot(drawingInfo.toXPx(point.left), drawingInfo.toYPx(point.top), dotRadius)
        }
        context.strokeStyle = strokeColor.colorCode
        context.lineWidth = 2.0
        context.fillStyle = fillColor.colorCode
        context.imageSmoothingEnabled = true
        context.stroke(dotPath)
        context.fill(dotPath)
        context.imageSmoothingEnabled = false
    }

    private fun Path2D.addDot(xPx: Double, yPx: Double, dotRadius: Double) {
        moveTo(xPx, yPx)
        arc(xPx, yPx, dotRadius, 0.0, FULL_CIRCLE_ARG)
    }

    fun getInteractionPoint(pointPx: Point): InteractionPoint? {
        for (bound in interactionBounds.reversed()) {
            val closePoint = bound.interactionPoints.lastOrNull { it.isAround(pointPx) }
            if (closePoint != null) {
                return closePoint
            }
        }

        return null
    }

    private fun InteractionPoint.isAround(pointPx: Point): Boolean {
        val leftPx = drawingInfo.toXPx(left)
        val topPx = drawingInfo.toYPx(top)
        return abs(leftPx - pointPx.left) < 6 && abs(topPx - pointPx.top) < 6
    }

    companion object {
        private const val DOT_RADIUS = 3.2
        private const val FULL_CIRCLE_ARG = 2 * PI
    }
}

File: monosketch-svelte/src/lib/mono/workspace/canvas/interaction-canvas-view-controller.ts
import type { Point } from "$libs/graphics-geo/point";
import {
    type InteractionBound,
    LineInteractionBound,
    ScalableInteractionBound,
} from "$mono/shape-interaction-bound/interaction-bound";
import type { InteractionPoint } from "$mono/shape-interaction-bound/interaction-point";
import type { AppUiStateManager } from "$mono/ui-state-manager/app-ui-state-manager";
import { BaseCanvasViewController } from '$mono/workspace/canvas/base-canvas-controller';
import { type ThemeColor, ThemeColors } from '$mono/ui-state-manager/states';

const DOT_RADIUS = 3.2;
const SMALL_DOT_RADIUS = 1.0;
const FULL_CIRCLE_ARG = 2 * Math.PI;

/**
 * A class for managing the interaction canvas.
 * This class is responsible for drawing interaction bounds and the interaction indicators on the
 * canvas.
 */
export class InteractionCanvasViewController extends BaseCanvasViewController {
    private interactionBounds: InteractionBound[] = [];
    private isMouseMoving = false;

    constructor(
        canvas: HTMLCanvasElement,
        private appUiStateManager: AppUiStateManager,
    ) {
        super(canvas);
        this.context.imageSmoothingQuality = 'high';
    }

    setInteractionBounds = (interactionBounds: InteractionBound[]) => {
        this.interactionBounds = interactionBounds;
    };

    setMouseMoving = (isMouseMoving: boolean) => {
        this.isMouseMoving = isMouseMoving;
    };

    protected drawInternal = () => {
        for (const bound of this.interactionBounds) {
            if (bound instanceof LineInteractionBound) {
                this.drawLineInteractionBound(bound);
            } else if (bound instanceof ScalableInteractionBound) {
                this.drawScalableInteractionBound(bound);
            }
        }
    };

    private drawLineInteractionBound = (bound: LineInteractionBound) => {
        if (this.isMouseMoving) {
            return;
        }

        this.drawInteractionPoints(
            bound.interactionPoints,
            ThemeColors.SelectionBoundStroke,
            ThemeColors.SelectionDotFill,
            DOT_RADIUS,
        );

        // Draw a small dot inside the start dot to make it simper to distinguish start and end dots
        this.drawInteractionPoints(
            [bound.interactionPoints[0]],
            ThemeColors.SelectionBoundStroke,
            ThemeColors.SelectionDotStroke,
            SMALL_DOT_RADIUS,
        );
    };

    private drawScalableInteractionBound = (bound: ScalableInteractionBound) => {
        const context = this.context;
        context.strokeStyle = this.appUiStateManager.getThemedColorCode(
            ThemeColors.SelectionBoundStroke,
        );
        context.lineWidth = 1.0;
        context.stroke(this.createScalableBoundPath(bound));

        if (this.isMouseMoving) {
            return;
        }
        this.drawInteractionPoints(
            bound.interactionPoints,
            ThemeColors.SelectionBoundStroke,
            ThemeColors.SelectionDotFill,
            DOT_RADIUS,
        );
    };

    private drawInteractionPoints = (
        points: InteractionPoint[],
        strokeColor: ThemeColor,
        fillColor: ThemeColor,
        dotRadius: number,
    ) => {
        const context = this.context;
        const drawingInfo = this.drawingInfo;
        const dotPath = new Path2D();
        context.beginPath();
        for (const point of points) {
            const xPx = drawingInfo.toXPx(point.left);
            const yPx = drawingInfo.toYPx(point.top);
            dotPath.moveTo(xPx, yPx);
            dotPath.arc(xPx, yPx, dotRadius, 0, FULL_CIRCLE_ARG);
        }
        context.strokeStyle = this.appUiStateManager.getThemedColorCode(strokeColor);
        context.fillStyle = this.appUiStateManager.getThemedColorCode(fillColor);
        context.imageSmoothingEnabled = true;
        context.stroke(dotPath);
        context.fill(dotPath);
        context.imageSmoothingEnabled = false;
        context.closePath();
    };

    private createScalableBoundPath = (bound: ScalableInteractionBound): Path2D => {
        const drawingInfo = this.drawingInfo;
        const leftPx = drawingInfo.toXPx(bound.left);
        const rightPx = drawingInfo.toXPx(bound.right);
        const topPx = drawingInfo.toYPx(bound.top);
        const bottomPx = drawingInfo.toYPx(bound.bottom);
        const path = new Path2D();
        path.rect(leftPx, topPx, rightPx - leftPx, bottomPx - topPx);
        return path;
    };

    getInteractionPoint = (pointPx: Point): InteractionPoint | null => {
        for (const bound of this.interactionBounds.reverse()) {
            const interactionPoint = this.getClosePointOfBound(bound, pointPx);
            if (interactionPoint) {
                return interactionPoint;
            }
        }
        return null;
    };

    private getClosePointOfBound(bound: InteractionBound, pointPx: Point) {
        for (const interactionPoint of bound.interactionPoints.reverse()) {
            const leftPx = this.drawingInfo.toXPx(interactionPoint.left);
            const topPx = this.drawingInfo.toYPx(interactionPoint.top);
            if (Math.abs(leftPx - pointPx.left) < 6 && Math.abs(topPx - pointPx.top) < 6) {
                return interactionPoint;
            }
        }
    }
}
